auth/password_hash_test.go

```diff
@@ -64,3 +67,394 @@ func TestSetBcryptCost(t *testing.T) {
+	require.NoError(b, err)
+	return hash
+}
+
+// hashAndPassword represents the password hash pair.
+type hashAndPassword struct {
+	hash     []byte // bcrypt hashed password
+	password []byte // plaintext equivalent
+}
+
+// generateTestData generates the test data and warm up the cache if required.
+func generateTestData(b *testing.B, cache Cache, warmCache bool, key string, numPasswords int) []hashAndPassword {
+	hashAndPasswords := make([]hashAndPassword, numPasswords)
+	for i := 0; i < numPasswords; i++ {
+		password := []byte(fmt.Sprintf("%s%d", key, i))
+		hashAndPassword := hashAndPassword{
+			hash:     bcryptHash(b, password),
+			password: password,
+		}
+		hashAndPasswords[i] = hashAndPassword
+		if warmCache {
+			key := authKey(hashAndPassword.hash, hashAndPassword.password)
+			cache.Put(key)
+		}
+	}
+	return hashAndPasswords
+}
+
+func BenchmarkCompareHashAndPassword100PercentCacheHit(b *testing.B) {
+	const maxCacheSize = 100
+	benchmarks := []struct {
+		name  string
+		cache Cache
+	}{
+		{"RandReplKeyCache", NewRandReplKeyCache(maxCacheSize)},
+		{"NoReplKeyCache", NewNoReplKeyCache(maxCacheSize)},
+	}
+	for _, bm := range benchmarks {
+		b.Run(bm.name, func(b *testing.B) {
+			b.ReportAllocs()
+			cache := bm.cache
+			defer cache.Purge()
+			testData := generateTestData(b, cache, true, "foo", maxCacheSize)
+
+			b.ResetTimer()
+			for i := 0; i < b.N; i++ {
+				hashAndPassword := testData[i%maxCacheSize]
+				_ = compareHashAndPassword(cache, hashAndPassword.hash, hashAndPassword.password)
+			}
+		})
+	}
+}
+
+func BenchmarkCompareHashAndPasswordCacheMissAndFill(b *testing.B) {
+	const maxCacheSize = 100
+	benchmarks := []struct {
+		name  string
+		cache Cache
+	}{
+		{"RandReplKeyCache", NewRandReplKeyCache(maxCacheSize)},
+		{"NoReplKeyCache", NewNoReplKeyCache(maxCacheSize)},
+	}
+	for _, bm := range benchmarks {
+		b.Run(bm.name, func(b *testing.B) {
+			b.ReportAllocs()
+			cache := bm.cache
+			defer cache.Purge()
+			testData := generateTestData(b, cache, false, "bar", maxCacheSize)
+
+			b.ResetTimer()
+			for i := 0; i < b.N; i++ {
+				hashAndPassword := testData[i%maxCacheSize]
+				_ = compareHashAndPassword(cache, hashAndPassword.hash, hashAndPassword.password)
+			}
+		})
+	}
+}
+
+func BenchmarkCompareHashAndPassword100PercentIncorrect(b *testing.B) {
+	const maxCacheSize = 100
+	benchmarks := []struct {
+		name  string
+		cache Cache
+	}{
+		{"RandReplKeyCache", NewRandReplKeyCache(maxCacheSize)},
+		{"NoReplKeyCache", NewNoReplKeyCache(maxCacheSize)},
+	}
+	for _, bm := range benchmarks {
+		b.Run(bm.name, func(b *testing.B) {
+			b.ReportAllocs()
+			cache := bm.cache
+			defer cache.Purge()
+			hash := bcryptHash(b, []byte("foo"))
+
+			b.ResetTimer()
+			for i := 0; i < b.N; i++ {
+				_ = compareHashAndPassword(cache, hash, []byte("baz"))
+			}
+		})
+	}
+}
+
+func BenchmarkPutAndContains(b *testing.B) {
+	const (
+		maxCacheSize = 100
```

Comment:
Increasing cache size to match our default is probably good for this test, to better measure the cost of Contains and Put.  

Running with numKeys = 2 x cache size seems reasonable.

Reply:
Sure, the cache size has been doubled while measuring the cost of Contains and Put. Here is the latest benchmark results.

```
BenchmarkPutAndContains/RandReplKeyCache-16         	 3417339	       363 ns/op	      16 B/op	       2 allocs/op
BenchmarkPutAndContains/NoReplKeyCache-16           	 4266928	       283 ns/op	      56 B/op	       2 allocs/op

Iteration: 0, Cache Name: RandReplKeyCache, Cache Hits: 1, Cache Misses: 0
Iteration: 1, Cache Name: RandReplKeyCache, Cache Hits: 34, Cache Misses: 66
Iteration: 2, Cache Name: RandReplKeyCache, Cache Hits: 3765, Cache Misses: 6235
Iteration: 3, Cache Name: RandReplKeyCache, Cache Hits: 499621, Cache Misses: 500379
Iteration: 4, Cache Name: RandReplKeyCache, Cache Hits: 1707270, Cache Misses: 1710069
Iteration: 5, Cache Name: NoReplKeyCache, Cache Hits: 1, Cache Misses: 0
Iteration: 6, Cache Name: NoReplKeyCache, Cache Hits: 37, Cache Misses: 63
Iteration: 7, Cache Name: NoReplKeyCache, Cache Hits: 3770, Cache Misses: 6230
Iteration: 8, Cache Name: NoReplKeyCache, Cache Hits: 278953, Cache Misses: 721047
Iteration: 9, Cache Name: NoReplKeyCache, Cache Hits: 1190307, Cache Misses: 3076621
PASS
```